### ISSUE TITLE
When drawing directed edges, account for node boundary width

### ISSUE THREAD

**Issue by jeremiastraub** [networkx/networkx#4601]
When drawing `nx.DiGraph`s with `arrows=True`, the `node_size` argument in `nx.draw_networkx_edges` may be used to ensure that the arrow heads end right at the node borders. However, the current default value (=300) as well as drawing via `nx.draw` or `nx.draw_networkx` don't take node boundaries (i.e. the marker edges of the scatter points) into account.

The width of the node boundaries is determined by the `linewidths` argument (can be turned off by setting `edgecolors` to 'none'), which defaults to `rcParams["lines.linewidth"]` (default=1.5).

When increasing the line width, the marker edge covers the arrow head:

```python
g = nx.DiGraph([[0, 1], [1, 2], [2, 0]])

nx.draw(g, linewidths=[7, 0, 3], edgecolors="r")
```
yields

![networkx_issue_patch_boundary](https://user-images.githubusercontent.com/60784028/107024172-facefa80-67a7-11eb-9f7e-295332fbd3da.png)

But also when using the defaults, a tiny bit of the arrow head is covered by the marker edge:

```python
g = nx.DiGraph([[0, 1], [1, 2], [2, 0]])

nx.draw(g)
```
yields

![networkx_issue_patch_boundary_basic](https://user-images.githubusercontent.com/60784028/107023964-a9267000-67a7-11eb-9c54-bac28d30880f.png)

### Proposal

If I see it correctly, doing something like `+ linewidths / 2` to `shrink_source` and `shrink_target` (as parsed in `nx.draw_networkx_edges`) should fix this.

I would propose to add `linewidths` and `edgecolors` as `nx.draw_networkx_edges` arguments. Then use them (together with `node_size`) to calculate better `shrink_source` and `shrink_target` values. (One would also have to handle the cases where `linewidths` and `edgecolors` are iterables.)

I see that there is a note saying:
```python
# Note: Waiting for someone to implement arrow to intersection with
# marker.  Meanwhile, this works well for polygons with more than 4
# sides and circle.
```

However, I think that any more sophisticated arrow-to-intersection method would also need the `linewidths` and `edgecolors` information, so the proposal above could be a useful intermediate step.

**Comment by rossbar** [comment]
My feeling is that there probably isn't a good solution to this problem currently. As you point out above with the linked code comment, the current arrow/marker intersection is inexact, so even if you pass in the linewidths and get things tweaked so they work well with one marker style, chances are the changes won't work well with another marker style.

For example, if I run your original example with a different marker:

```python
 nx.draw(g, linewidths=[7, 0, 3], edgecolors="r", node_shape="^")
```

![Figure_1](https://user-images.githubusercontent.com/1268991/107103432-75ad0b00-67d2-11eb-92f5-e439f0c5b3e0.png)

you see the figure has the same problem, but in a different way! This all goes to say that I'm not sure the problem can truly be resolved until it's possible to compute line/marker intersections.

One thing you can try that may or may not be adequate for your use-case is to use the `min_source_margin` and `min_target_margin` keywords to explicitly specify a standoff distance between edges and nodes:

```python
nx.draw(g, linewidths=[7, 0, 3], edgecolors="r", min_source_margin=25, min_target_margin=25)
```

![Figure_2](https://user-images.githubusercontent.com/1268991/107103707-6c706e00-67d3-11eb-9aca-390ce23ca40e.png)

At least this trick can be used to avoid the arrowhead/marker overlap problem and tends to make the differing edge lengths less noticeable.

**Comment by jeremiastraub** [comment]
> This all goes to say that I'm not sure the problem can truly be resolved until it's possible to compute line/marker intersections.

I agree that the proposal does not address the different marker types. But perhaps accounting for the node boundaries can be addressed separately. Surely, when eventually calculating the *exact* edge/node intersection the node-shape and boundary linewidth have to be accounted for *together*, as you illustrated it. But for the current (partially workaround) implementation, it might be enough to just add the extra space for the node boundaries, without any surprising behavior when changing the marker type. What I mean to say is that the inexactness for the different marker types would remain as it is (should be checked systematically), but node boundaries could be handled (as opposed to now).

When adding `linewidths / 2` to `shrink_source` and `shrink_target`...

... it is exact for circular markers,
![comparison-circles](https://user-images.githubusercontent.com/60784028/107119362-5623ea00-6887-11eb-89dc-3924b722c1fb.png)

... it works well for polygons with more than 4 sides,
![comparison-hexagons](https://user-images.githubusercontent.com/60784028/107119365-6045e880-6887-11eb-8b27-7223541a9486.png)

... and it increases the extra-space given to large markers such that the arrow heads are shown
![comparison-triangles](https://user-images.githubusercontent.com/60784028/107119368-63d96f80-6887-11eb-8f16-866bc4267c52.png)

Of course, it might also be a question of whether it's worth improving the inexact solution and extending the workaround, if an exact solution might be implemented soon?

**Comment by rossbar** [comment]
> Of course, it might also be a question of whether it's worth improving the inexact solution and extending the workaround, if an exact solution might be implemented soon?

I'm not entirely sure what's currently possible with matplotlib or whether anyone's working on an intersection-based solution for NX (there are currently no open PRs for it at least) so my sense is that it's not on the immediate horizon. In that case, maybe it's worth extending the workaround - your examples above clear do a better job handling the case where the nodes have large edge widths at least.

**Comment by jeremiastraub** [comment]
Agree, I'm not even sure whether a comprehensive intersection-based solution will be feasible at all, given the variety of [possible markers](https://matplotlib.org/stable/api/markers_api.html) – particularly since any (custom) `Path` instance may be used as marker...

> In that case, maybe it's worth extending the workaround

I'll open a PR once I've found the time to do this.

**Comment by rossbar** [comment]
Gentle ping @jeremiastraub - any updates on this?

**Comment by jeremiastraub** [comment]
thanks for the reminder @rossbar. Didn't get around to do it yet, and it will be another 3-4 weeks until I have the time. But I'll be happy to take up the task then.
